driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java

```diff
@@ -102,37 +107,43 @@
             this.operationTime = result.getTimestamp(OPERATION_TIME, null);
             this.postBatchResumeToken = getPostBatchResumeTokenFromResponse(result);
         }
-        if (firstQueryResult.getCursor() != null) {
+        ServerCursor serverCursor = initFromQueryResult(firstQueryResult);
+        if (serverCursor != null) {
             notNull("connectionSource", connectionSource);
         }
-        if (connectionSource != null) {
-            this.connectionSource = connectionSource.retain();
-        }
-
-        initFromQueryResult(firstQueryResult);
         firstBatchEmpty = firstQueryResult.getResults().isEmpty();
-
+        Connection connectionToPin = null;
+        boolean releaseServerAndResources = false;
         if (connection != null) {
             this.maxWireVersion = connection.getDescription().getMaxWireVersion();
             if (limitReached()) {
-                killCursor(connection);
+                releaseServerAndResources = true;
             } else {
                 assertNotNull(connectionSource);
                 if (connectionSource.getServerDescription().getType() == ServerType.LOAD_BALANCER) {
-                    this.connection = connection.retain();
-                    this.connection.markAsPinned(Connection.PinningMode.CURSOR);
+                    connectionToPin = connection;
                 }
             }
         }
-        releaseConnectionAndSourceIfNoServerCursor();
+        closeableState = new CloseableState(namespace, connectionSource, connectionToPin, serverCursor);
```

Comment:
I think we could do with a better name for this class / variable. To reflect its use - essentially its the manager of the cursor / connection etc.

Reply:
Renamed to `ResourceManager`.